src/array_api_stubs/_draft/info.py

```diff
@@ -0,0 +1,149 @@
+__all__ = [
+    "__array_namespace_info__",
+    "capabilities",
+    "default_device",
+    "default_dtypes",
+    "devices",
+    "dtypes",
+]
+
+from ._types import (
+    Optional,
+    Union,
+    Tuple,
+    List,
+    device,
+    dtype,
+    DefaultDataTypes,
+    DataTypes,
+    Capabilities,
+    Info,
+)
+
+
+def __array_namespace_info__() -> Info:
+    """
+    Returns a namespace with Array API namespace inspection utilities.
+
+    Returns
+    -------
+    out: Info
+        Namespace with Array API namespace inspection utilities.
+    """
+
+
+def capabilities() -> Capabilities:
+    """
+    Returns a dictionary of array library capabilities.
+
+    The dictionary should contain the following keys:
+
+    -   **boolean_indexing**: boolean indicating whether an array library supports boolean indexing.
+    -   **data_dependent_shapes**: boolean indicating whether an array library supports data-dependent output shapes.
+
+    Returns
+    -------
+    out: Capabilities
+        a dictionary of array library capabilities.
+    """
+
+
+def default_device() -> device:
+    """
+    Returns the default device.
+
+    Returns
+    -------
+    out: device
+        an object corresponding to the default device.
+    """
+
+
+def default_dtypes(
+    *,
+    device: Optional[device] = None,
+) -> DefaultDataTypes:
+    """
+    Returns a dictionary containing default data types.
+
+    The dictionary should have the following keys:
+
+    -   **real floating**: default real floating-point data type.
+    -   **complex floating**: default complex floating-point data type.
+    -   **integral**: default integral data type.
+    -   **indexing**: default index data type.
```

Comment:
I don't recall how this one is supposed to work, e.g., why would users care, which API does this impact, etc?

Reply:
This API provides a means of programmatically querying the default data types for the different defaults described in the specification. E.g., for a given dtype "kind", this API would allow downstream libraries to check if the default dtype is of a lower (or higher) precision than desired, or what the default dtype is for a specific device (e.g., on CPU, this could be "float64", but on GPU, this could be "float32", etc). When working with a library supporting multiple device types, a downstream library wanting a specific minimum precision could then choose a device based on which device automatically supports operations having the desired default device type.

Reply:
Thanks for the explanation @kgryte. I guess I shouldn't have been vague. I understand the purpose of default data types (int/float), they all sound good. My specific question is on the `indexing` one, which I am clueless how knowing this would help (ex: which API mentions default indexing type?), and how a compliant library should report, which there's dynamic mode switching (ex: CuPy can decide on the fly the indexing type used in a JIT'd kernel)?

Reply:
@leofang We discuss the default array index data type in the following locations:

- Data Types: https://data-apis.org/array-api/2022.12/API_specification/data_types.html#default-data-types
- argmax: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.argmax.html#array_api.argmax
- argmin: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.argmin.html#array_api.argmin
- nonzero: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.nonzero.html#array_api.nonzero
- argsort: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.argsort.html#array_api.argsort
- unique\*: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.unique_all.html#array_api.unique_all

In short, wherever we return indices, we state that the default return data type be the default array index data type.

As for dynamic mode switching, I don't think we have considered this. But wouldn't this apply similarly, in theory, to all default data types? Couldn't a library choose to dynamically switch the "default" floating-point data type used in a JIT'd kernel?

Reply:
But also, correct me if I am wrong, dynamic mode switching seems more like an implementation detail, whereas this API concerns the default array type used for indexing into an array in Array API consuming code.

Additionally, in, e.g., `take`, I can imagine Array API consumers wanting to allocate the `indices` array according to the respective array index data type to avoid unnecessary casts. Having a way to query that data type would help ensure alignment.

Am I missing something here?

Reply:
Thanks, @kgryte, I guess I totally overlooked (or forgot about) the indexing data type. It makes sense for the `arg*` or any index-returning APIs to consider.

> But wouldn't this apply similarly, in theory, to all default data types? Couldn't a library choose to dynamically switch the "default" floating-point data type used in a JIT'd kernel?

Yes, in theory. However, I would think it's a lot easier to fix the default int/float data types, and only vary the indexing data type dynamically based on the input size (which offers a natural condition for run-time dispatching). Though, in practice it seems NumPy/CuPy both do have the notion of a default indexing data type, so my job should be easy and I won't complain 😄
```py
>>> import numpy as np
>>> a = np.random.random(2**15)
>>> b = a.argsort()
>>> b.dtype  # it's int64, despite the problem fits in int32
dtype('int64')
```

Reply:
Sorry, didn't noticed your 2nd reply until I hit send. No, I think we're good here. I resolved this thread.

Reply:
Thanks, @leofang!